### PR TITLE
Fix/mtp效率和稳定性优化

### PR DESCRIPTION
**Pull request by jklnvlink**
## PR1 摘要

> **问题 #722（死锁）**：双卡 TP（偶数 TP）+ MTP 长上下文 prefill 永久卡死。偶数 TP 的集合通信对大张量会回退 NCCL；一个 rank 在 peer 仍在提交 NCCL 时进入下一个 GEMM/分配器（持 CUDA 驱动锁并隐式同步），双方在驱动锁上互等 → 死锁。实测卡死点恰在 256 页分页参数边界。原代码仅 odd-TP 创建 `NcclSubmitRendezvous`（`numGPUs % 2 != 0`），偶数 TP 缺这道会合。
>
> **修复 A（`fastllm-multicuda.cu`）**：把会合创建条件从 `numGPUs % 2 != 0` 放宽为 `numGPUs > 1`——所有多 rank TP 组在 NCCL 提交前会合。`NcclSubmitRendezvous` 类为上游既有代码（`include/devices/multicuda/ncclsubmitrendezvous.h`），本修复不新增该类。
>
> **修复 B（`fastllm-attention.cu` + `cudadevice.cpp`）**：分页参数（qSizes/pageSizes/pageIndexs/lastPageLens）原先经 4 次 pageable H2D memcpy 上传，DMA 期间持 CUDA 驱动锁，正是死锁窗口之一。改为把参数值放进 kernel 参数空间由单线程 kernel 写入显存：无阻塞拷贝、可被 CUDA Graph 捕获；超过 kernel 参数预算（sm_70+ 为 32 KiB，最大档 MaxPages=4096 约 17 KiB）时返回 false，调用方保留原 memcpy 回退，行为不变。
>
> **问题 #724（设备 fault）**：custom all-reduce kernel 以 16 字节对齐包访问 data/dest，收到未对齐指针时设备侧 fault（`cudaErrorMisalignedAddress` / Xid 13）。另：custom AR 仅靠指针元组按到达顺序配对 rank，任一 rank 的集合通信调用流发生失步（例如某 rank 多调/漏调一次），后续每一轮都会配错 peer 张量——静默算错或越界读已释放的 peer buffer。
>
> **修复 C（`fastllm-custom-allreduce.cu`）**：① 三个候选函数（`RunCustomArCandidate`/`RunCustomArPairAddCandidate`/`RunCustomArAddCandidate`）入口加 16B 对齐守卫，未对齐返回 false 回退 NCCL；② `CustomArState` 增加 per-rank 调用序号（`rankCallSequence`/`pendingSequences`），到齐后比对各 rank 序号，失步则该轮拒绝（回退 NCCL）并一次性打日志；③ `FastllmCudaCustomAllReduceReset` 复位全部序号状态。
>
> 三处修复均为纯防御性：不改变任何数值路径；失败一律回退既有 NCCL/memcpy 路径。

## PR2 摘要
上游 #728
> **现状（`4b0f717d`）**：`2f457292` 已修掉 MTP 采样分布偏差，手段是**保留 one-hot 贪心草稿 + 精确接受**：草稿 `d = argmax`（源码注释明言 "The MTP head proposes greedy (one-hot) drafts."，`qwen3_5.cpp:17054`；单请求/非 TP 走 `sampleGreedyFromCudaLogits` `:30570`，TP 走 `RunMtpTpDraft` per-rank `FastllmCudaGreedySamplingWithScores` `:29544`），接受判据 `speculativeMtpAccepted[row] = sampled[row] == mtpDraftIds[i]`（`:17069`）。数学上这等价于**链拒绝 + one-hot 草稿分布 `q = δ_argmax`**：输出分布精确无偏，但单位置接受率上限被钉死在 `p(argmax)`——长思考高熵段 `p(argmax)` 仅 0.4–0.7，`2f457292` commit message 自报吞吐 **−16.2%**。
>
> **本 PR（S3）**：补上缺的一步——**把草稿从 one-hot 分布化**。每个草稿位在与验证侧**完全相同**的过滤链（temperature softmax → top-K renorm → top-P renorm，复用 `4b0f717d` 验证侧同一组 kernel）内做 Gumbel-max 采样得草稿 token `d`，并收集该位 top-K 候选；整条草稿链结束后把 `drafts + 逐位候选` 发布进既有的 `DFlashContext.proposal*`。验证前向复用 `4b0f717d` **保留下来的** `FastllmCudaDFlashRejectionSampling` 链拒绝分支（kernel 零改动）：`y == d` → 接受；`y ∈ 候选` → scatter 接受 `y`、链继续；否则 → 提交 `y` 残差、断链。
>
> **正确性**：链拒绝理论保证输出分布**精确等于目标 `p`**（Leviathan et al. 2023，与 `2f457292` 同级正确性）。
> **接受率**：单位置接受率从 `p(argmax)` 变为 `overlap(p,q) = Σ_v min(p(v), q(v))`，恒有 `overlap(p,q) ≥ min(p(argmax), q(argmax)) + Σ_{v≠argmax} min(p(v), q(v))`；当 `q` 为 `p` 的良近似（`q(argmax) ≥ p(argmax)` 且 `∃v≠argmax: p(v)q(v) > 0`）时**严格大于 `p(argmax)`**——S3 草稿与目标同过滤空间、MTP head 近似目标，条件成立。`2f457292` 的 one-hot 草稿即 `q = δ_argmax` 退化端点（第二项恒 0，接受率恰为 `p(argmax)`），故 S3 在良近似条件下严格优于现状。
>
> **启用条件**：单请求 && 非贪心（`generationConfigs.size() == 1 && !IsSimpleGreedy()`）时自动启用，无 env 开关；其余情形（贪心、多请求批处理、DFlash 权重模型）走原路径，代码路径不变。注意 `4b0f717d` 的 `Qwen35MtpSupportsGenerationConfig` 对 `repeat_penalty ≠ 1` 整体禁用 MTP，此类请求本就无 MTP 草稿，不受影响。
> **候选宽度**：`K = clamp(请求 top_k, 1, 64)`，与验证侧 `max(1, top_k)` 钳位一致。
> **回退**：git revert（本 PR 独立可摘；无运行时开关）。
>
> **实测（fork 同源实现，base 早于 `2f457292`，2×2080Ti / Qwen3.8-27B-FP8 / TP=2，非贪心 T=1.0/top_p=0.95/top_k=20）**：三位接受率 **89.37 / 77.24 / 66.17**（S3 前同模型基线 73.2/54.0/41.5），tok/step **3.328（+23.9%**），形态保持几何衰减（无准贪心锁定回归），32k–169k 长 prefill 播种 19 次全存活，0 Xid / 0 Assert。**在 `4b0f717d` 上的 A/B 重测为本 PR 落地前置验证**。

## PR2 机制图

```
草稿侧（每个草稿位一次，单请求且非贪心时启用）
  MTP head logits ──→ temperature softmax ──→ top-K renorm ──→ top-P renorm
      （与验证侧 target 完全相同的过滤链 → 草稿 q 与目标 p 同处一个过滤空间）
      ──→ Gumbel-max 采样 → 草稿 token d
          （score = log(p) + log(u)；u 由 splitmix64(seed, index) 无状态派生，
            seed = 首个草稿步 steady_clock 时间戳 + 每草稿位自增计数器，可复现）
      ──→ 收集该位 top-K（p>0）候选（ids + 归一化概率）
  整条草稿链结束：drafts + 逐位候选 → 发布到 dflashContext.proposal*
  落点：替换 4b0f717d 的 RunMtpGreedyDraft 非 TP argmax(:30570)、
        RunMtpTpDraft per-rank argmax 合并(:29577-:29600)、
        RunMtpGreedyDraft 自有 TP argmax 合并(:30641-:30673)

验证侧
  目标前向对扩展序列打分 → 目标采样 y ~ p（4b0f717d 既有采样，未改）
      复用 4b0f717d 保留的 FastllmCudaDFlashRejectionSampling(:16997) 做链拒绝：
        y == d           → 接受草稿 d（链继续）
        y ∈ 候选列表      → scatter 接受 y（链继续，提交 y）
        否则             → 提交 y 残差，断链
  落点：runTargetWithPast 的 speculativeDFlashSamplingContext 路由(:18166-:18168)
        + DFlash 拒绝分支候选宽度(:16944-:17003 处 effSelectorTopK)
        + countAcceptedDrafts 接受计数(:18275-:18277)
```

## 不足
1.PR1的具体影响机制仍缺乏直接证据，因此均采取防御性修复，如此一来可能会造成不必要的代码修改

### CODE REVIEW

**Comment by jklnvlink** [comment]
[mtp.md](https://github.com/user-attachments/files/32159735/mtp.md)

## 实验配置

| 配置项 | 值 |
|--------|----|
| 模型 | Qwen3.8-27B-FP8 + FP8 KV Cache |
| GPU | 2× RTX 2080 Ti 22GB (NVLink 双 lane) |
| MTP 层数 | 1 |
| Drafts per step | 4 |
| 协议 | OpenAI Chat Protocol |
| PR1 (Prompt Routing 1) | 开启 |
| PR2 (Prompt Routing 2) | 对比关闭 / 开启 |

> **注意**：本实验无 PR1=OFF PR2=OFF 基线数据，下述对比均为 **PR1=ON 固定前提下**，PR2 OFF vs ON 的差异。额外增加无偏修复后的对比数据。

## Commit 4b0f717合并实测
在默认pr1开启的情况下，以不同情景测试pr2开启或者关闭时mtp接受率
1.**纯对话场景，代码任务**：
| 指标 | PR2=OFF | PR2=ON | 变化 |
|------|---------|--------|------|
| 平均速度 | 73.2 t/s | 78.3 t/s | **+6.9%** |
| 标准差 | 16.8 | 17.9 | — |
| 峰值速度 | 105.3 t/s | 108.0 t/s | — |
| 最低速度 | 38.8 t/s | 38.8 t/s | — |

2.**cladue工具调用、代码审查、总结对话**：
| 指标 | PR2=OFF (119k prefill) | PR2=ON (135k prefill) | 变化 |
|------|------------------------|----------------------|------|
| 平均速度 | 49.8 t/s | 57.6 t/s | **+15.7%** |
| 中位数 | 48.5 t/s | 56.8 t/s | **+17.1%** |
| 标准差 | 12.1 | 12.5 | — |
| 峰值速度 | 81.3 t/s | 79.7 t/s | — |
| 最低速度 | 28.6 t/s | 31.5 t/s | — |

3.**claude实际使用场景中mtp接受率对比**：
| Position | PR2=OFF | PR2=ON | 提升幅度 |
|----------|---------|--------|----------|
| pos 1（首个 draft token） | 71% | **87%** | **+16%** |
| pos 2 | 52% | **72%** | **+20%** |
| pos 3 | 37% | **58%** | **+21%** |
| pos 4（末位 draft） | 28% | **48%** | **+20%** |

4.**claude实际使用场景中prefill性能对比**：
| 配置 | Prefill 总 token 数 | 收尾速度 | 备注 |
|------|-------------------|----------|------|
| PR2=OFF | 119,042 | ~520 t/s | 速度渐进下降 |
| PR2=ON | 134,948 | ~587 t/s | 速度更稳定 |

5.**claude实际使用场景中decode波动性对比**：
| 配置 | CV | 含义 |
|------|-----|------|
| PR2=OFF | 0.244 | 基线 |
| PR2=ON | **0.216** | **降低 11.3%** |

**Comment by jklnvlink** [comment]
相较于dflash2，mtp的主要提升在于128k上下文以后的平均decode速度，以及对内存更低的需求，低内存设备（64g以下）使用起来会更加稳定
